### ISSUE TITLE
Prompt caching support

### ISSUE THREAD

**Issue by voberoi** [pydantic/pydantic-ai#138]
Is there prompt caching support? I couldn't find evidence in the docs.

If not, would love to have that baked in.

**Comment by samuelcolvin** [comment]
Happy to add support if we can find a good way to make it work in a model/API-agnostic way.

I tried using Gemini's caching but it doesn't work unless the context is massive.

**Comment by Appilot-dev** [comment]
Andriod karma farming bot that runs on your device controlled from your dashboard.

We have made a reddit bot in 2024 but it functions through your phone remotely with a web dashboard no ADB or wire needed, a reddit karma bot that mimics human behavior through a learning script for the bot that shadows different humanized ways of managing an account of reddit.
You can check it's commenting features in [our repository](https://github.com/Appilot-dev/Reddit-Android-Karma-Farming-Bot)
How you can control the bot’s speed and never get banned again.

**Comment by shdmitry2000** [comment]
> Happy to add support if we can find a good way to make it work in a model/API-agnostic way.
> 
> I tried using Gemini's caching but it doesn't work unless the context is massive.

gemini work with cache only above 2000 tokens. clade work from 500  if I remember well.

**Comment by nikolaidk** [comment]
Not having it is the only thing stopping me from using framework: makes it way to expense for my usecases.

**Comment by roberthoenig** [comment]
I guess you could set the `http_client` on your `Model` instance to use a cache-enabled httpx client like Hishel (https://hishel.com/)?

Something like 
```.py
model = OpenAIModel('gpt-4o', http_client=hishel.AsyncCacheClient())
```
(Haven't tried this out yet.)

**Comment by karpetrosyan** [comment]
Can help with integration

**Comment by mrcwinn** [comment]
+1 Would love to see this. In our case we use `cache_control` checkpoint system through Anthropic's API. So I hear you on "model agnostic" but perhaps another way is flexibility into the prompt injection itself, as cache checkpoints are tied to the message or tool, not the overall model chat call.

**Comment by voberoi** [comment]
Returning to this a few months in, after:

- Using pydantic-ai extensively and enjoying it.
- Needing Claude's models for some tasks, because they are very well-suited to it.
- Needing prompt-caching to make things economical, where I am re-prompting ~100K+ tokens at a time.

I have also poked around the code and can't see a clear path to making this model-agnostic.

I reached out to Anthropic to see if they'll be supporting implicit prefix-caching like OpenAI, vs. their explicit cache-control message. They don't seem to be prioritizing this.

I'd love to be able to use this feature with pydantic-ai. I'm maintaining/writing a lot more code than I need to, for these use cases.

**Comment by voberoi** [comment]
Would be open to discussing a PR. I'm not entirely sure how this should look because as @mrcwinn describes, it breaks the model abstraction in pydantic-ai a bit.

**Comment by karpetrosyan** [comment]
Such code worked for me. The idea is that by injecting the HTTP client into the model and changing the underlying HTTP transport communication, we can cache HTTP responses. So, if we have already made a request and the client finds the response to that particular request in its cache, it won’t send any new requests; instead, it will use the cached response.

```python
from hishel._utils import generate_key
from pydantic_ai import Agent
from pydantic_ai.models.groq import GroqModel
import httpx
import hishel
import httpcore


def custom_key_generator(request: httpcore.Request, body: bytes):
    key = generate_key(request, body)
    method = request.method.decode()
    host = request.url.host.decode()
    return f"{method}|{host}|{key}"


cache_client = httpx.AsyncClient(
    transport=hishel.AsyncCacheTransport(
        transport=httpx.AsyncHTTPTransport(),
        storage=hishel.AsyncFileStorage(),
        controller=hishel.Controller(
            force_cache=True,
            cacheable_methods=["GET", "POST"],
            key_generator=custom_key_generator,
        ),
    )
)

agent = Agent(
    GroqModel(
        "llama-3.3-70b-versatile",
        http_client=cache_client,
        api_key="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",
    ),
    system_prompt="Be concise, reply with one sentence.",
)

result = agent.run_sync('Where does "hello" come from?')
```

After making a new request that is not yet cached, a new file for that particular cache entry will be created. This is how it should appear in your filesystem:

![Image](https://github.com/user-attachments/assets/a060ee84-2a3e-440f-af70-08b012b66d7d)

Note that this is similar to force_caching. It means you understand and intentionally ignore any caching rules that come with the response, reusing it even if it was not originally designed for caching.

**Comment by voberoi** [comment]
@karpetrosyan that caches model responses, which is not what this issue concerns.

This issue is about enabling prompt caching on the provider end to reduce cost and latency.

In particular, I'm looking to handle prompt caching with Claude, which uses a "cache_control" setting in messages sent to the model: https://docs.anthropic.com/en/docs/build-with-claude/prompt-caching.

**Comment by voberoi** [comment]
Noting how this is done across the major frontier model providers so it's all in one place.

**Anthropic**

Anthropic requires explicit caching of a prompt prefix by inserting `"cache_control": {"type": "ephemeral"}` at the end of the prefix's contents in a regular model call: https://docs.anthropic.com/en/docs/build-with-claude/prompt-caching

Changes are required in what is allowed in messages/how messages are assembled in Pydantic AI.

Requested in #1041.

**Google**

Google requires explicit caching by using a separate call to cache context. This creates a name for the content.

That name needs to be used in a message you send to Google when requesting a completion: https://cloud.google.com/vertex-ai/generative-ai/docs/context-cache/context-cache-overview

I believe changes are required in what is allowed in messages/how messages are assembled in Pydantic AI.

Requested in #664.

**OpenAI**

Handles prompt caching transparently with matching prompt prefixes: //platform.openai.com/docs/guides/prompt-caching

No changes needed in Pydantic AI.

**Comment by tianshangwuyun** [comment]
Indeed, this is the biggest obstacle to using pydanticAI, and there is no good way to extend the

**Comment by timesler** [comment]
+1. I'm wanting to use pydantic-ai for software development agents, which would often see a 50-70% cost saving using Anthropic's prompt caching (for the agents I'm using at least). Also results in a significant, though more modest, reduction in latency. Support for prompt caching in message blocks would be a pretty massive win

**Comment by soichisumi** [comment]
+1
This is necessary to use pydantic-ai in production.
Not only system prompt, but also caching of tool definition is important
https://docs.anthropic.com/en/docs/build-with-claude/prompt-caching#prompt-caching-examples

**Comment by charlesmindee** [comment]
Hello +1 on this one

**Comment by danlester** [comment]
I'd appreciate caching fully on the client side (i.e. just don't make the HTTP call to the LLM if we have a previous result for the same inputs). Not based on the LLM provider's own caching layer. This is mainly for speed and cost savings, to be used in development and production.

Langchain has a great caching layer using Sqlite or similar data stores.

**Comment by DouweM** [comment]
Note that Gemini now supports implicit prompt caching: https://developers.googleblog.com/en/gemini-2-5-models-now-support-implicit-caching/

Like with OpenAI (https://platform.openai.com/docs/guides/prompt-caching), this means no specific action in PydanticAI is necessary to benefit fromthis.

**Comment by voberoi** [comment]
@danlester this issue relates to prompt caching support from providers to reduce cost and latency, not client side caching for e.g. testing.

There may be another issue open for client side caching.

@DouweM that's awesome. There's still explicit caching for Google, I think, if you want explicit guarantees. But I agree, there's less of a need to support ut now.

I think Anthropic is all that's left. I did reach out to their support a couple of weeks ago and they said implicit prompt caching was not on the roadmap.

**Comment by danlester** [comment]
OK I think #1093 relates to client-side caching. Thanks all.

**Comment by DouweM** [comment]
Closing in favor of https://github.com/pydantic/pydantic-ai/issues/664 (Google) and https://github.com/pydantic/pydantic-ai/issues/1041 (Anthropic). (I've reopened the latter)
